cpp/src/arrow/compare.cc

```diff
@@ -113,8 +113,9 @@ inline bool FloatingApproxEquals(const NumericArray<ArrowType>& left,
       return (fabs(x - y) <= epsilon) || (std::isnan(x) && std::isnan(y));
     });
   } else {
-    return BaseFloatingEquals<ArrowType>(
-        left, right, [epsilon](T x, T y) -> bool { return fabs(x - y) <= epsilon; });
+    return BaseFloatingEquals<ArrowType>(left, right, [epsilon](T x, T y) -> bool {
+      return (fabs(x - y) <= epsilon) || (x == y);
+    });
```

Comment:
You must also change the `nans_equal` path above.

Reply:
Revised accordingly. Thanks a lot for your kind reminder.